certbot-nginx/certbot_nginx/configurator.py

```diff
@@ -233,20 +238,46 @@ def choose_vhost(self, target_name):
         matches = self._get_ranked_matches(target_name)
         vhost = self._select_best_name_match(matches)
         if not vhost:
-            # No matches. Raise a misconfiguration error.
-            raise errors.MisconfigurationError(
-                        ("Cannot find a VirtualHost matching domain %s. "
-                         "In order for Certbot to correctly perform the challenge "
-                         "please add a corresponding server_name directive to your "
-                         "nginx configuration: "
-                         "https://nginx.org/en/docs/http/server_names.html") % (target_name))
+            if raise_if_no_match:
+                # No matches. Raise a misconfiguration error.
+                raise errors.MisconfigurationError(
+                            ("Cannot find a VirtualHost matching domain %s. "
+                             "In order for Certbot to correctly perform the challenge "
+                             "please add a corresponding server_name directive to your "
+                             "nginx configuration: "
+                             "https://nginx.org/en/docs/http/server_names.html") % (target_name))
+            else:
+                return None
         else:
             # Note: if we are enhancing with ocsp, vhost should already be ssl.
             if not vhost.ssl:
                 self._make_server_ssl(vhost)
 
         return vhost
 
+    def _vhost_from_duplicated_default(self, domain):
+        if self.new_vhost is None:
+            default_vhost = self._get_default_vhost()
+            self.new_vhost = self.parser.create_new_vhost_from_default(default_vhost)
+            if not self.new_vhost.ssl:
+                self._make_server_ssl(self.new_vhost)
+            self.new_vhost.names = set()
+
+        self.new_vhost.names.add(domain)
+        name_block = [['\n    ', 'server_name', ' ', " ".join(self.new_vhost.names)]]
+        self.parser.add_server_directives(self.new_vhost, name_block, replace=True)
+        return self.new_vhost
+
+    def _get_default_vhost(self):
+        vhost_list = self.parser.get_vhosts()
+        # if one has default_server set, return that one
+        for vhost in vhost_list:
+            for addr in vhost.addrs:
+                if addr.default:
+                    return vhost
```

Comment:
Unfortunately, I'm not sure it's that simple. When Nginx determines which server block will handle a request, it looks at the IP address and port (or UNIX socket) the request came in on, the server name, and the default server for that IP and port being used. Here we're ignoring the IP and port.

For example, imagine your Nginx config has the following three server blocks:
```
server {
    listen 80 default_server;
    server_name ipv4.example.org;
    root /var/www/html/ipv4;
}

server {
    listen 127.0.0.1:80 default_server;
    server_name localhost.example.org;
    root /var/www/html/localhost;
}

server {
    listen [::]:80 default_server;
    server_name ipv6.example.org;
    root /var/www/html/ipv6;
}
```
When using this PR, which server block is copied? The one that happens to come first in your config.

This won't matter in the majority of cases, but I think it will matter in some more complex setups. While you could argue that maybe we shouldn't worry about handling these more complicated configs, I think we should detect the problem and inform the user rather than trying to use an arbitrary server block.

We won't know the IP and port that will be used by Nginx when you try to connect to a domain, so I think we can only do this automatic selection in the case where there's a single choice and one server block is the default for every address found in the config. (Perhaps we could ignore the port or only look at port 80/443, but I'm not sure if that's always accurate either.) In all other cases, I think we need more information from the user.